packages/marshal/src/encodePassable.js

```diff
@@ -241,32 +244,44 @@ const encodeArray = (array, encodePassable) => {
 /**
  * @param {string} encoded
  * @param {(encoded: string) => Passable} decodePassable
+ * @param {number} [skip]
  * @returns {Array}
  */
-const decodeArray = (encoded, decodePassable) => {
-  encoded.startsWith('[') || Fail`Encoded array expected: ${encoded}`;
+const decodeArray = (encoded, decodePassable, skip = 0) => {
   const elements = [];
   const elemChars = [];
-  for (let i = 1; i < encoded.length; i += 1) {
-    const c = encoded[i];
-    if (c === '\u0000') {
+  // Use a string iterator to avoid slow indexed access in XS.
+  // https://github.com/endojs/endo/issues/1984
+  let stillToSkip = skip + 1;
+  let inEscape = false;
+  for (const c of encoded) {
+    if (stillToSkip > 0) {
+      stillToSkip -= 1;
+      if (stillToSkip === 0) {
+        c === '[' || Fail`Encoded array expected: ${encoded.slice(skip)}`;
+      }
+    } else if (inEscape) {
+      c === '\u0000' ||
+        c === '\u0001' ||
+        Fail`Unexpected character after u0001 escape: ${c}`;
+      elemChars.push(c);
+    } else if (c === '\u0000') {
       const encodedElement = elemChars.join('');
       elemChars.length = 0;
       const element = decodePassable(encodedElement);
       elements.push(element);
     } else if (c === '\u0001') {
-      i += 1;
-      i < encoded.length || Fail`unexpected end of encoding ${encoded}`;
-      const c2 = encoded[i];
-      c2 === '\u0000' ||
-        c2 === '\u0001' ||
-        Fail`Unexpected character after u0001 escape: ${c2}`;
-      elemChars.push(c2);
+      inEscape = true;
+      // eslint-disable-next-line no-continue
```

Comment:
Not a comment on this PR, but curious: Do you think we should turn off warning on `continue`? I don't see the value in it.

Reply:
I would _love_ to disable that warning; continue is as useful as `break` and early `return` and I don't see any value in discouraging it.